### DISCUSSION TITLE
What's the difference between ClipExtension and MaskExtension？

### DISCUSSION THREAD

**Discussion by Awen-hub** [visgl/deck.gl#7392] [Q&A] [answered]
I want to render the partial layer which is in the viewable area on the current screen. It seems ClipExtension and MaskExtension can be used to do this. So what's their difference and which one should I use?

**Comment by AdriSolid** [comment, marked as answer]
Basically:
- `ClipExtension`: allows clipping rendered layers by rectangular bounds on the layer itself
- `MaskExtension`: allows layers to show/hide objects by a geofence (here you will need a layer that acts as a geofence to tell the target layer what features should be included or excluded)

In your case, `ClipExtension` makes more sense, as always your mask will be the viewport (rectangular). And you don't need to create other layer as a mask:
```javascript
const layer = new YourLayer({
  ...,
  extensions: [new ClipExtension()],
  clipBounds: [-180, -85.051129, 180, 85.051129], // viewport: west, south, east, north
});
```

**Reply by Awen-hub** [reply]
Thanks for your answer. I also wonder whether the ClipExtension will greatly improve the render performance. Is it the main purpose of this extension?

**Reply by kylebarron** [reply]
I think the main point of clip extension was to correctly render Mapbox Vector Tiles, which have overlaps between each tile

**Reply by AdriSolid** [reply]
At some point `ClipExtension` was located in `geo-layers/mvt-layer` folder, and then moved to `extensions` as it could be really helpful for clipping other layers based on rectangles. As it is excluding features on the GPU, it can improve the performance.

**Reply by Awen-hub** [reply]
```
import { LayerExtension, _ShaderModule as ShaderModule } from "@deck.gl/core";

import type { Layer } from "@deck.gl/core";
import { ClipExtensionProps } from "@deck.gl/extensions/typed";

const shaderFunction = `
uniform vec4 clip_bounds;

bool clip_isInBounds(vec2 position) {
  return position.x >= clip_bounds[0] && position.y >= clip_bounds[1] && position.x < clip_bounds[2] && position.y < clip_bounds[3];
}
`;

/*
 * The vertex-shader version clips geometries by their anchor position
 * e.g. ScatterplotLayer - show if the center of a circle is within bounds
 */
const shaderModuleVs: ShaderModule = {
  name: "clip-vs",
  vs: shaderFunction,
};

const injectionVs = {
  "vs:#decl": `
varying float clip_isVisible;
`,
  "vs:DECKGL_FILTER_GL_POSITION": `
  clip_isVisible = float(clip_isInBounds(geometry.worldPosition.xy));
`,
  "fs:#decl": `
varying float clip_isVisible;
`,
  "fs:DECKGL_FILTER_COLOR": `
  if (clip_isVisible < 0.5) discard;
`,
};

/*
 * The fragment-shader version clips pixels at the bounds
 * e.g. PolygonLayer - show the part of the polygon that intersect with the bounds
 */
const shaderModuleFs: ShaderModule = {
  name: "clip-fs",
  fs: shaderFunction,
};

const injectionFs = {
  "vs:#decl": `
varying vec2 clip_commonPosition;
`,
  "vs:DECKGL_FILTER_GL_POSITION": `
  clip_commonPosition = geometry.position.xy;
`,
  "fs:#decl": `
varying vec2 clip_commonPosition;
`,
  "fs:DECKGL_FILTER_COLOR": `
  if (!clip_isInBounds(clip_commonPosition)) discard;
`,
};

/** Adds support for clipping rendered layers by rectangular bounds. */
export default class ClipWithViewportExtension extends LayerExtension {
  static extensionName = "ClipWithViewportExtension";

  getShaders(this: Layer<ClipExtensionProps>) {
    let clipByInstance =
      "instancePositions" in this.getAttributeManager().attributes;
    // Users can override by setting the `clipByInstance` prop
    if ("clipByInstance" in this.props) {
      clipByInstance = this.props.clipByInstance as boolean;
    }
    this.state.clipByInstance = clipByInstance;

    return clipByInstance
      ? {
          modules: [shaderModuleVs],
          inject: injectionVs,
        }
      : {
          modules: [shaderModuleFs],
          inject: injectionFs,
        };
  }

  /* eslint-disable camelcase */
  draw(this: Layer<ClipExtensionProps>, { uniforms, context }: any): void {
    const clipBounds = context.viewport.getBounds();
    if (this.state.clipByInstance) {
      uniforms.clip_bounds = clipBounds;
    } else {
      const corner0 = this.projectPosition([clipBounds[0], clipBounds[1], 0]);
      const corner1 = this.projectPosition([clipBounds[2], clipBounds[3], 0]);

      uniforms.clip_bounds = [
        Math.min(corner0[0], corner1[0]),
        Math.min(corner0[1], corner1[1]),
        Math.max(corner0[0], corner1[0]),
        Math.max(corner0[1], corner1[1]),
      ];
    }
  }
}
```

I have wrote an extension named ClipWithViewportExtension just like ClipExtension. It will clip uniforms by `viewport.getBounds()`. However, I can't see an obvious improvement for performance in fps when rendering a large dataset. So, I think ClipExtension can not improve performance at most time.

**Reply by Awen-hub** [reply]
I just use it with polygonLayer and scatterplotLayer. Maybe tilelayer will get an obvious improvement.

**Reply by Pessimistress** [reply]
The purpose of either extension is to render your data with an additional effect. They do not make rendering faster. If anything, they add overhead to your render procedure.

The only way to improve performance of a very big dataset is to load and render less data. In which case you probably want the TileLayer.

**Reply by Awen-hub** [reply]
`While dependent on zoom levels (clipping will improve performance to some extent) this many fragments will certainly strain even a recent MacBook Pro GPU.` was said in docs so I think ClipExtension may help with performance. Maybe, `to some extent` is right.

**Reply by Awen-hub** [reply]
Is there any to make GPU ignore the datas which are out of viewport boundary? Maybe it will help.

**Reply by Pessimistress** [reply]
Geometries outside of the viewport do not get sent to the fragment shader; you may want to read about the [rendering pipeline](https://www.khronos.org/opengl/wiki/Rendering_Pipeline_Overview).

TileLayer only draws data that are visible in the current viewport and zoom level.

**Reply by Awen-hub** [reply]
> Geometries outside of the viewport do not get sent to the fragment shader; you may want to read about the [rendering pipeline](https://www.khronos.org/opengl/wiki/Rendering_Pipeline_Overview).
> 
> TileLayer only draws data that are visible in the current viewport and zoom level.

Thanks, I get it.
